### DISCUSSION TITLE
(Atari) Unexpected effect of -O on PIA behaviour

### DISCUSSION THREAD

**Discussion by 8bit-Dude** [cc65/cc65#2739] [General]
I have connected an arduino to joystick port 2 of Atari 8bit, and written a piece of code for sending/receiving bits.
After spending 2 days trying to understanding why the PIA did not output the correct bits, I removed (by chance) the -O compile options, and suddenly the PIA started outputting correct bits.

I have attached the C code below, as well as the S file listing without/with the -O compile option. 
[hub-test.zip](https://github.com/cc65/cc65/files/7690981/hub-test.zip)

**Comment by mrdudz** [comment]
I dont know the PIA - but you should generally be careful with programm any I/O chips directly in C - because many chips have registers that, for example, are cleared once read, or where a certain condition is triggered when they have been written. The compiler doesnt know about this, and might eg read a register twice that clears on read. Or the compiler might generate instructions that cause "dummy accesses" (like indexed instructions do) and then the code will read or write at locations that were totally not intended. The bottom line ist: this kind of code should always be written in assembler

**Comment by polluks** [comment]
...and we need volatile!

**Comment by mrdudz** [comment]
volatile will not help much with this problem actually - the compiler would still generate eg indexed instructions which cause all kind of unexpected behaviour

**Comment by 8bit-Dude** [comment]
Thanks for the tip MRDUDZ. I will probably write that part of the code directly in ASM to avoid unexpected behaviour.

**Comment by greg-king5** [comment]
You can make that function return `value` directly, instead of going indirectly through an argument.  Also, declare `value` as `static`.  Those changes should avoid indirect indexing instructions in that function..

**Comment by 8bit-Dude** [comment]
Making value static did not solve the problem. Also, I want to return whether the Send/Recv was successful (thus the use of value).
